src/anomalib/models/efficientad/torch_model.py

```diff
@@ -0,0 +1,350 @@
+        aug_index = random.choice([1, 2, 3])
+        # Sample an augmentation coefficient λ from the uniform distribution U(0.8, 1.2)
+        coefficient = random.uniform(0.8, 1.2)
+        if aug_index == 1:
+            img_aug = transforms.functional.adjust_brightness(image, coefficient)
+        elif aug_index == 2:
+            img_aug = transforms.functional.adjust_contrast(image, coefficient)
+        elif aug_index == 3:
+            img_aug = transforms.functional.adjust_saturation(image, coefficient)
+        return img_aug
+
+    def forward(self, batch: Tensor, batch_imagenet: Tensor = None) -> Union[Tensor, dict]:
+        """Prediction by EfficientAD models.
+
+        Args:
+            batch (Tensor): Input images.
+
+        Returns:
+            Tensor: Predictions
+        """
+        with torch.no_grad():
+            teacher_output = self.teacher(batch)
+            if self.is_set(self.mean_std):
+                teacher_output = (teacher_output - self.mean_std["mean"]) / self.mean_std["std"]
+            _, c, h, w = teacher_output.shape
+
+        student_output = self.student(batch)
+        ae_output = self.ae(batch)
+        # 3: Split the student output into Y ST ∈ R 384×64×64 and Y STAE ∈ R 384×64×64 as above
+        student_output = student_output[:, : self.teacher_out_channels, :, :]
+        student_output_ae = student_output[:, -self.teacher_out_channels :, :, :]
+
+        distance_st = torch.pow(teacher_output - student_output, 2)
+
+        if self.training:
+            # Student loss
+            d_hard = torch.quantile(distance_st, 0.999)
+            loss_hard = torch.mean(distance_st[distance_st >= d_hard])
+            student_imagenet_output = self.student(batch_imagenet)[:, :self.teacher_out_channels, :, :]
+            loss_st = loss_hard + (1 / (c * h * w)) * torch.sum(torch.pow(student_imagenet_output, 2))
+
+            # Autoencoder and Student AE Loss
+            aug_img = self.choose_random_aug_image(batch)
+            ae_output_aug = self.ae(aug_img)
+            student_output_aug = self.student(aug_img)
+            student_output_ae_aug = student_output_aug[:, -self.teacher_out_channels :, :, :]
+
+            with torch.no_grad():
+                teacher_output_aug = self.teacher(aug_img)
+                if self.is_set(self.mean_std):
+                    teacher_output_aug = (teacher_output_aug - self.mean_std["mean"]) / self.mean_std["std"]
+
+            distance_ae = torch.pow(teacher_output_aug - ae_output_aug, 2)
+            distance_stae = torch.pow(ae_output_aug - student_output_ae_aug, 2)
+
+            loss_ae = torch.mean(distance_ae)
+            loss_stae = torch.mean(distance_stae)
+
+            return (loss_st, loss_ae, loss_stae)
+
+        else:
+            distance_stae = torch.pow(ae_output - student_output_ae, 2)
+
+            map_st = torch.mean(distance_st, dim=1, keepdim=True)
+            map_stae = torch.mean(distance_stae, dim=1, keepdim=True)
+
```

Comment:
For accurate segmentation, maps need to be padded again after this, see https://github.com/nelson1425/EfficientAD/blob/a44786ccddd3697b54903394ee1361f38e65de16/efficientad.py#L204

Reply:
Alright it's now added

Reply:
When reproducing the paper the padding of the output was necessary for the segmentation metrics to be the same. Without the padding of the output the segmentation metrics were very bad. I think it is because if we do not pad the output then the resizing of the anomaly map will displace some of the anomaly scores. Similar to when we run an edge filter with a filter size of 11x11 across a 100x100 image. This will output a 90x90 image. If we resize the 90x90 image to 100x100 again then the edge detections will be displaced. Instead we need to pad the image with 5 pixels on each border.